x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyAuthenticator.java

```diff
@@ -60,7 +60,7 @@ public void authenticate(Context context, ActionListener<AuthenticationResult<Au
                     ? authResult.getException()
                     : Exceptions.authenticationError(authResult.getMessage());
                 logger.debug(() -> "API key service terminated authentication for request [" + context.getRequest() + "]", e);
-                // TODO: emit audit event
+                context.getRequest().exceptionProcessingRequest(e, authenticationToken);
                 listener.onFailure(e);
```

Comment:
This means we will also audit whenever the thread pool is saturated (i.e., we're getting 429s). It could mean that the audit log gets swamped with more events, in case someone is hitting the endpoint with a lot of requests.

The alternative I see here is auditing _only_ "authentication expected API key type" type failures (we could even properly wrap these in 401s potentially), either by inspecting the error message or adding a cause [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L1116), and inspecting that.

Overall though, I'm on board with your approach: it's simpler, and, usually, having audit events for the 429s can actually be useful. I don't think the extra audit log noise in case of many requests should be that big of a concern. WDYT?

Reply:
In principle I think we should audit all failure events. I consider it a bug that we don't audit anything when API key authentication is terminated. Today a 429 API key authentication leaves no trace at all in the audit log which feels off. So my intention is to audit all terminations including both 429 and type mis-match.

> It could mean that the audit log gets swamped with more events, in case someone is hitting the endpoint with a lot of requests.

If there are a lot of requests, there will be a lot events anyway. If all these requests succeed, they will generate `access_granted` events. If they fail to authorize, they will generate `access_denied`. So I think it is reasonble to have `authentication_failed` events if they fail to authenticate for any reasons including 429 (in fact, they already do for realm users).